### PR TITLE
feat: Implement device login handler

### LINKED ISSUE #432: Implement Login Handler for device login rest api

### PR DESCRIPTION
**Pull request by talwinder50**
closes #432

Signed-off-by: talwinder.kaur <talwinder.kaur@securekey.com>

### CODE REVIEW

File: pkg/restapi/device/operations.go [hunk 1]
```diff
@@ -188,22 +188,98 @@ func (o *Operation) finishRegistration(w http.ResponseWriter, r *http.Request) {
...
 	logger.Debugf("Registration success: %s", o.uiEndpoint)
 }
 
+func (o *Operation) beginLogin(w http.ResponseWriter, r *http.Request) {
+	logger.Debugf("handling begin device login: %s", r.URL.String())
+
+	// get username
+	username := r.URL.Query().Get("username")
+	if username == "" {
+		common.WriteErrorResponsef(w, logger, http.StatusBadRequest, "missing useername parameter")
```
Comment (sandrask) [inline pkg/restapi/device/operations.go:212]: username instead of useername

File: pkg/restapi/device/operations_test.go [hunk 2]
```diff
@@ -302,20 +317,201 @@ func TestRegistration_FinishRegistration(t *testing.T) {
...
+		w := httptest.NewRecorder()
+
+		c.beginLogin(w, newDeviceLoginRequest(userSub))
+		require.Equal(t, http.StatusBadRequest, w.Code)
+	})
+	t.Run("begin login - missing username parameter", func(t *testing.T) {
+		w := httptest.NewRecorder()
+
+		o.beginLogin(w, newDeviceRegistrationRequest())
+		require.Equal(t, http.StatusBadRequest, w.Code)
```
Comment (sandrask) [inline pkg/restapi/device/operations_test.go:380]: You should consider checking for specific errors in your tests. If you change this code later on and your code now erroneously returns top level error all the time all of these tests will still pass. In this case you can catch coding issues only if you inspect code coverage results - unit tests will not catch it.

**Review by sandrask — APPROVED** [review]

**Review by rolsonquadras — APPROVED** [review]

File: pkg/restapi/device/operations_test.go (hunk 2 above)
Comment (talwinder50) [reply pkg/restapi/device/operations_test.go:382]: done

File: pkg/restapi/device/operations.go [hunk 3]
```diff
@@ -188,22 +188,98 @@ func (o *Operation) finishRegistration(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	credentialBytes, err := marshallResponse(credential)
+	device.AddCredential(*credential)
```
Comment (llorllale) [inline pkg/restapi/device/operations.go:191]: For us, there's no sense in having an `AddCredential()` method on the device (see https://github.com/trustbloc/edge-agent/pull/417#discussion_r502611775)

File: pkg/restapi/device/operations.go [hunk 4]
```diff
@@ -188,22 +188,98 @@ func (o *Operation) finishRegistration(w http.ResponseWriter, r *http.Request) {
...
+	jsonResponse(w, credential, http.StatusOK)
 
 	logger.Debugf("Registration success: %s", o.uiEndpoint)
 }
 
+func (o *Operation) beginLogin(w http.ResponseWriter, r *http.Request) {
+	logger.Debugf("handling begin device login: %s", r.URL.String())
+
+	// get username
+	username := r.URL.Query().Get("username")
```
Comment (llorllale) [inline pkg/restapi/device/operations.go:210]: why isn't this grabbed from a cookie?

File: pkg/restapi/device/operations.go [hunk 5]
```diff
@@ -231,6 +307,50 @@ func (o *Operation) getUserData(w http.ResponseWriter, r *http.Request) (userDat
...
 
-func marshallResponse(value interface{}) ([]byte, error) {
-	return json.Marshal(value)
+func (o *Operation) saveDeviceInfo(device *Device) error {
+	deviceBytes, err := json.Marshal(device)
+	if err != nil {
+		return fmt.Errorf("failed to marshall the device data %s", err)
+	}
+
+	err = o.store.transient.Put(device.ID, deviceBytes)
```
Comment (llorllale) [inline pkg/restapi/device/operations.go:316]: device info shouldn't be saved just to transient storage

**Review by llorllale — CHANGES_REQUESTED** [review]
See my comments

File: pkg/restapi/device/operations.go (hunk 3 above)
Comment (talwinder50) [reply pkg/restapi/device/operations.go:191]: begin and finish login endpoints duo-lab library expects the credential, without that it doesnt work. I can discuss offline if it doesnt make sense

Comment (llorllale) [reply pkg/restapi/device/operations.go:191]: @talwinder50 as per our discussion:

Agreed, let's leave the 1:N device:credential relationships until we understand the use cases better (such as the device registering multiple key types in different credentials).

TODO - understand these use cases better

File: pkg/restapi/device/operations.go (hunk 4 above)
Comment (talwinder50) [reply pkg/restapi/device/operations.go]: grabbed from cookie now

File: pkg/restapi/device/operations.go (hunk 5 above)
Comment (talwinder50) [reply pkg/restapi/device/operations.go]: done moved to permanent storage

**Comment by llorllale** [comment]
@talwinder50 the latest changes look good. 

Can you increase the coverage a bit in order to merge?

**Review by llorllale — APPROVED** [review]

**Pull request merged**